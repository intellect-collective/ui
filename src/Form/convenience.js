import { asField } from '../utils';

const Hidden = asField('value', { type: 'hidden' })('input');
const NumberField = asField('value', { type: 'number' })('input');
const Text = asField('value', { type: 'text' })('input');
const Textarea = asField('value')('textarea');
const Password = asField('value', { type: 'password' })('input');

export {
    Hidden,
    NumberField,
    Password,
    Text,
    Textarea
};
