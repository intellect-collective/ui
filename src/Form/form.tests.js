import React from 'react';
import { Form } from '../..';

describe('Form', () => {
    it('Missing onSubmit does not produce error', () => {
        const wrapper = mount(
            <Form action="/">
                <input type="submit" />
            </Form>
        );
        wrapper.find('input').simulate('submit');
    });

    it('Form submit calls handler', () => {
        const fn = jest.fn();
        const wrapper = mount(
            <Form action="/" handler={{ onSubmit: fn }}>
                <input type="submit" name="test" />
            </Form>
        );
        wrapper.find('input').simulate('submit');
        expect(fn).toHaveBeenCalledTimes(1);
        expect(fn.mock.calls[0][0]).toMatchObject({
            type: 'submit',
            target: {
                name: 'test'
            }
        });
    });

    it('Can set form encoding type', () => {
        const text = mount(<Form action="/" encType="text" />);
        expect(text.find('form').prop('encType')).toEqual('text/plain');
        const multipart = mount(<Form action="/" encType="multipart" />);
        expect(multipart.find('form').prop('encType')).toEqual('multipart/form-data');
    });

    it('Can set method to post', () => {
        const wrapper = mount(
            <Form action="/" method="post" />
        );
        expect(wrapper.find('form').prop('method')).toEqual('post');
    });

    it('Can set method to put', () => {
        const wrapper = mount(
            <Form action="/" method="put" overrideProperty="__method__" />
        );
        expect(wrapper.find('form').prop('method')).toEqual('post');
        expect(wrapper.find('input').prop('name')).toEqual('__method__');
        expect(wrapper.find('input').prop('value')).toEqual('put');
    });

    it('Can render component', () => {
        const fn = jest.fn(() => (<div />));
        mount(
            <Form action="/" component={ fn } />
        );
        expect(fn).toHaveBeenCalledTimes(1);
        expect(Object.keys(fn.mock.calls[0][0])).toIncludeSameMembers([
            'children',
            'getOriginal',
            'getLastValue',
            'getValue',
            'setValue',
            'getError',
            'setError',
            'isTouched',
            'setTouched',
            'isDirty',
            'getConflict'
        ]);
    });

    it('Can render function', () => {
        const fn = jest.fn(() => (<div />));
        mount(
            <Form action="/">{ fn }</Form>
        );
        expect(fn).toHaveBeenCalledTimes(1);
        expect(Object.keys(fn.mock.calls[0][0])).toIncludeSameMembers([
            'getOriginal',
            'getLastValue',
            'getValue',
            'setValue',
            'getError',
            'setError',
            'isTouched',
            'setTouched',
            'isDirty',
            'getConflict'
        ]);
    });
});
