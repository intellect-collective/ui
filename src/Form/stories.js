import React from 'react';
import PropTypes from 'prop-types';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withState } from '@dump247/storybook-state';
import {
    Checkbox, Field, Form, Password, Text
} from '../..';

class FormExample extends React.Component {
    static propTypes = {
        data: PropTypes.object,
        onChange: PropTypes.func
    };

    render() {
        return (
            <Form action="/" method="post" data={ this.props.data } onChange={ this.props.onChange } onSubmit={ (data, ev) => { ev.preventDefault(); ev.stopPropagation(); action('onSubmit')(data); } } className="form-horizontal">
                <p>Sign in for a better experience.</p>
                <Field name="email"
                        label="Email"
                        placeholder="Email"
                        className="form-group"
                        classNames={{
                            label: 'col-lg-2 control-label',
                            wrapper: 'col-lg-10',
                            field: 'form-control'
                        }}
                        wrapper="div"
                        component={ Text } />
                <Field name="password"
                        label="Password"
                        placeholder="Password"
                        className="form-group"
                        classNames={{
                            label: 'col-lg-2 control-label',
                            wrapper: 'col-lg-10',
                            field: 'form-control'
                        }}
                        wrapper="div"
                        component={ Password } />
                <div className="form-group">
                    <div className="col-lg-offset-2 col-lg-10">
                        <div className="i-checks">
                            <label>
                                <div className="icheckbox_square-green" style={{ position: 'relative' }}>
                                    <input type="checkbox" style={{ position: 'absolute', opacity: '0' }} />
                                    <ins className="iCheck-helper"
                                            style={{
                                                position: 'absolute',
                                                top: '0%',
                                                left: '0%',
                                                display: 'block',
                                                width: '100%',
                                                height: '100%',
                                                margin: '0px',
                                                padding: '0px',
                                                background: 'rgb(255, 255, 255)',
                                                border: '0px',
                                                opacity: '0'
                                            }} />
                                </div>
                                <i />
                                Remember me
                                <Checkbox name="remember" value="yes" />
                            </label>
                        </div>
                    </div>
                </div>
                <div className="form-group">
                    <div className="col-lg-offset-2 col-lg-10">
                        <button type="submit" name="go" className="btn btn-sm btn-white">Sign In</button>
                    </div>
                </div>
            </Form>
        );
    }
}

storiesOf('2 - Components / Form', module)
    .add('default', withState()(({ store }) => (<FormExample action="/" data={ store.state } onChange={ (ev) => store.set({ [ev.target.name]: ev.target.value}) } />)));
