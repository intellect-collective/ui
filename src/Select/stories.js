import React from 'react';
import { storiesOf } from '@storybook/react';
import { Select } from '../..';
import width from '../../.storybook/decorators/width';
import stateful from '../../.storybook/decorators/stateful';

storiesOf('Select', module)
    .addDecorator(stateful())
    .addDecorator(width(200))
    .add('default', () => (onChange, state) => (
        <Select name="actors"
                value={ state.actors }
                onChange={ (ev) => { onChange(ev.target.name, ev.target.value); } }>
            <option value="jackie">Jackie Chan</option>
            <option value="jet">Jet Li</option>
            <option value="tony">Tony Jaa</option>
        </Select>
    ))
    .add('multiple', () => (onChange, state) => (
        <Select name="actors"
                value={ state.actors }
                onChange={ (ev, value) => { onChange(ev.target.name, value || ev.target.value); } }
                multiple>
            <option value="jackie">Jackie Chan</option>
            <option value="jet">Jet Li</option>
            <option value="tony">Tony Jaa</option>
        </Select>
    ));
