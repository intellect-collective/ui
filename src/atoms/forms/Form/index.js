import React from 'react';
import PropTypes from 'prop-types';

export const FormContext = React.createContext(null);

export default class Form extends React.Component {
    static propTypes = {
        action: PropTypes.string.isRequired,
        method: PropTypes.oneOf(['get', 'post', 'put']),
        overrideProperty: PropTypes.string,
        encType: PropTypes.oneOf(['text', 'multipart']),
        onChange: PropTypes.func,
        onSubmit: PropTypes.func,
        children: PropTypes.any,
        data: PropTypes.object
    };

    static defaultProps = {
        method: 'post',
        overrideProperty: '_method',
        data: {},
        onChange: () => {},
        onSubmit: () => {}
    };

    constructor(props) {
        super(props);
        this.getValue = this.getValue.bind(this);
        this.setValue = this.setValue.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.getEncType = this.getEncType.bind(this);
        this.getFormMethod = this.getFormMethod.bind(this);
        this.getMethodOverrideElement = this.getMethodOverrideElement.bind(this);
    }


    getValue(name) {
        if (this.props.data) {
            return this.props.data[name];
        }
        return undefined;
    }

    setValue(ev) {
        this.props.onChange(ev);
    }

    /**
     * Calls the onSubmit prop function, if provided, when the form is
     * submitted.
     * @param {Object} e - The React event
     */
    onSubmit(e) {
        if (this.props.onSubmit) {
            this.props.onSubmit(this.props.data, e);
        }
    }

    /**
     * Determines the encoding type of the form.
     * See https://www.ietf.org/rfc/rfc1867.txt
     * @returns {String} - The mime type to specify in th enctype attribute
     */
    getEncType() {
        if (this.props.encType === 'text') {
            return 'text/plain';
        }
        if (this.props.encType === 'multipart') {
            return 'multipart/form-data';
        }
        return this.props.encType;
    }

    /**
     * Determines the HTTP method to use when submitting the form.
     * See https://www.ietf.org/rfc/rfc1867.txt
     * @returns {String} - The http method to use
     */
    getFormMethod() {
        if (['get', 'post'].indexOf(this.props.method) !== -1) {
            return this.props.method;
        }
        return 'post';
    }

    /**
     * Produces a hidden element which provides the opportunity to emulate
     * http methods other than GET and POST. Requires that the server receiving
     * the request support method overriding.
     * @returns {Object} - A React element
     */
    getMethodOverrideElement() {
        if (['get', 'post'].indexOf(this.props.method) !== -1) {
            return null;
        }
        return (
            <input type="hidden"
                    name={ this.props.overrideProperty || '_method' }
                    value={ this.props.method } />
        );
    }

    /**
     * Builds the props that are to be passed through to the form element. We
     * need to get rid of some props that aren't supported by the form element,
     * so we take that opportunity here.
     * @returns {Object} - The final props object
     */
    getProps() {
        const props = {
            ...this.props,
            encType: this.getEncType(),
            method: this.getFormMethod(),
            onSubmit: this.onSubmit
        };
        delete props.overrideProperty;
        return props;
    }

    /**
     * Perform render
     * @returns {Object} - A React element
     */
    render() {
        const {
            children,
            onChange,
            onSubmit,
            ...props
        } = this.getProps();
        const ctx = {
            getValue: this.getValue,
            setValue: this.setValue
        };
        return (
            <form { ...props }>
                <FormContext.Provider value={ ctx }>
                    { this.getMethodOverrideElement() }
                    { this.props.children }
                </FormContext.Provider>
            </form>
        );
    }
}
