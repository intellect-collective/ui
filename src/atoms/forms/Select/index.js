import React from 'react';
import PropTypes from 'prop-types';
import changeable from '../_decorators/changeable';
import formContext from '../_decorators/formContext';

const normalizeOptions = (options, val, placeholder, required) => {
    let result = [];

    // If we received an array of options, they should already be formatted
    // correctly
    if (Array.isArray(options)) {
        result = options.map(({ title, value }, i) => ({
            title,
            value,
            key: i
        }));
    } else {
        result = Object.keys(options).map((opt, i) => {
            return {
                title: options[opt],
                value: opt,
                key: i
            };
        });
    }
    if (placeholder || !required) {
        result.unshift({
            title: placeholder || 'Select One',
            disabled: Boolean(required),
            key: '__default__'
        });
    }
    return result;
};
const getValue = (options, value) => {
    if (typeof value !== 'undefined') {
        return value;
    }
    if (Array.isArray(options)) {
        return options.filter((o) => (o.selected))
            .map((o) => (o.value));
    }
    return undefined;
};

const Select = ({
    options, value, placeholder, required, ...props
}) => (
    <select { ...props } required={ required } value={ getValue(options, value) }>
        {
            normalizeOptions(options, value, placeholder, required)
                .map(({ title, ...option }) => (
                    <option { ...option }>
                        { title }
                    </option>
                ))
        }
    </select>
);
Select.displayName = 'Select';
Select.propTypes = {
    options: PropTypes.oneOfType([
        PropTypes.object,
        PropTypes.arrayOf(
            PropTypes.shape({
                title: PropTypes.string,
                value: PropTypes.oneOfType([
                    PropTypes.object,
                    PropTypes.string,
                    PropTypes.number
                ]),
                selected: PropTypes.bool
            })
        )
    ]).isRequired,
    value: PropTypes.oneOfType([
        PropTypes.object,
        PropTypes.string,
        PropTypes.number,
        PropTypes.arrayOf(
            PropTypes.oneOfType([
                PropTypes.object,
                PropTypes.string,
                PropTypes.number
            ])
        )
    ]),
    placeholder: PropTypes.string,
    required: PropTypes.bool
};

export default formContext(changeable(Select));
