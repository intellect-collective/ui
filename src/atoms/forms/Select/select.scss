@if (null != index($ui-includes, 'form.select') and null == index($ui-excludes, 'form.select')) {
//	select:not(.browser-default) {
//	    @extend .input;
//	}
}