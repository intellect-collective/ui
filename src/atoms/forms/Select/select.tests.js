import React from 'react';
import Select from '.';
import Form from '../Form';
import {
    getMissingFieldError
} from '../../../../test/utils';

const options = {
    hotdogs: 'Hotdogs',
    sausages: 'Sausages',
    burgers: 'Burgers'
};

const validProps = {
    options,
    name: 'test',
    value: 'sausages',
    onChange: () => {},
    required: true
};

describe('Select', () => {
    describe('Basics', () => {
        it('Should throw an error on missing name', () => {
            expect(() => {
                mount(<Select />);
            }).toThrowError(getMissingFieldError('name'));
        });

        it('Should accept valid props', () => {
            const wrapper = mount(<Select { ...validProps } required />);
            expect(wrapper).toMatchSnapshot();
            expect(wrapper.find('select').prop('name')).toEqual('test');
            expect(wrapper.find('option').length).toEqual(3);
            expect(wrapper.find('option[value="sausages"]').getDOMNode().selected).toEqual(true);
        });

        it('Clicking without handler is no-op', () => {
            const wrapper = mount(<Select { ...validProps } />);
            wrapper.find('select').simulate('change');
        });

        it('Clicking with handler supplies name and event', () => {
            const fn = jest.fn();
            const wrapper = mount(<Select { ...validProps } onChange={ fn } />);
            wrapper.find('select').simulate('change');
            expect(fn).toHaveBeenCalledWith(expect.objectContaining({
                target: expect.objectContaining({
                    name: validProps.name,
                    value: validProps.value
                })
            }));
        });
    });

    describe('Form Context Usage', () => {
        it('Gets checked state from form', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'sausages' }}>
                    <Select name="test" options={ options } required />
                </Form>
            );
            expect(wrapper.find('select').prop('name')).toEqual('test');
            expect(wrapper.find('option').length).toEqual(3);
            expect(wrapper.find('option[value="sausages"]').getDOMNode().selected).toEqual(true);
        });

        it('Prefers local attributes over form attributes', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'sausages' }}>
                    <Select name="test" options={ options } value="burgers" required />
                </Form>
            );
            expect(wrapper.find('select').prop('name')).toEqual('test');
            expect(wrapper.find('option').length).toEqual(3);
            expect(wrapper.find('option[value="sausages"]').getDOMNode().selected).toEqual(false);
            expect(wrapper.find('option[value="burgers"]').getDOMNode().selected).toEqual(true);
        });

        it('Remains unchecked if form field has incorrect value', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'streetmeat' }}>
                    <Select name="test" options={ options } required />
                </Form>
            );
            expect(wrapper.find('select').prop('value')).toEqual('streetmeat');
        });

        it('Can check multiple', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: ['hotdogs', 'sausages'] }}>
                    <Select name="test" options={ options } multiple required />
                </Form>
            );
            expect(wrapper.find('option[value="sausages"]').getDOMNode().selected).toEqual(true);
            expect(wrapper.find('option[value="hotdogs"]').getDOMNode().selected).toEqual(true);
        });
    });
});
