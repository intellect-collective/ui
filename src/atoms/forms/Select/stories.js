import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Select from './index';
import width from '../../../../.storybook/decorators/width';

const options = [
    {
        title: 'Jason Statham',
        value: 'jason'
    },
    {
        title: 'Jackie Chan',
        value: 'jackie'
    },
    {
        title: 'Terry Crews',
        value: { a: 1, b: 'test' }
    }
];

storiesOf('Atoms / Forms / Select', module)
    .addDecorator(width(200))
    .add('default', () => (<Select name="test-select-1" options={ options } value="jackie" onChange={ action('onChange') } />))
    .add('required', () => (<Select name="test-select-1" options={ options } value="jackie" onChange={ action('onChange') } required />))
    .add('optional with placeholder', () => (<Select name="test-select-1" options={ options } onChange={ action('onChange') } value="jackie" placeholder="Please Select..." />))
    .add('required with placeholder', () => (<Select name="test-select-1" options={ options } onChange={ action('onChange') } value="jackie" placeholder="Please Select..." required />))
    .add('multiple', () => (<Select name="test-select-1" options={ options } onChange={ action('onChange') } value={ ['jackie', 'jason'] } placeholder="Please Select..." multiple required />));
