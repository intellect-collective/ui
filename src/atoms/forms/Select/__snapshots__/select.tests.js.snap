// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Basics Should accept valid props 1`] = `
<FormContext
  name="test"
  onChange={[Function]}
  options={
    Object {
      "burgers": "Burgers",
      "hotdogs": "Hotdogs",
      "sausages": "Sausages",
    }
  }
  required={true}
  value="sausages"
>
  <FormContextInner
    ctx={null}
    name="test"
    onChange={[Function]}
    options={
      Object {
        "burgers": "Burgers",
        "hotdogs": "Hotdogs",
        "sausages": "Sausages",
      }
    }
    required={true}
    value="sausages"
  >
    <Changeable
      name="test"
      onChange={[Function]}
      options={
        Object {
          "burgers": "Burgers",
          "hotdogs": "Hotdogs",
          "sausages": "Sausages",
        }
      }
      required={true}
      value="sausages"
    >
      <Select
        name="test"
        onChange={[Function]}
        options={
          Object {
            "burgers": "Burgers",
            "hotdogs": "Hotdogs",
            "sausages": "Sausages",
          }
        }
        required={true}
        value="sausages"
      >
        <select
          name="test"
          onChange={[Function]}
          required={true}
          value="sausages"
        >
          <option
            key="0"
            value="hotdogs"
          >
            Hotdogs
          </option>
          <option
            key="1"
            value="sausages"
          >
            Sausages
          </option>
          <option
            key="2"
            value="burgers"
          >
            Burgers
          </option>
        </select>
      </Select>
    </Changeable>
  </FormContextInner>
</FormContext>
`;
